### ISSUE TITLE
Updating CUDA and CuPy improved memory management 

### ISSUE THREAD

**Issue by GearoidM** [jacobjma/abTEM#37]
Dear abTEM users, 

I thought it might be useful to share some knowledge I learned through trial and error with using abTEM. I hope this is an appropriate place for this message. For reference, I present an outline of what I am trying to achieve with abTEM. For convenience, I run 4DSTEM simulations on multiple structures in a loop. To elaborate: 

- I have a folder full of structure files (e.g. PDB or CIF format)
- My script loads the first file in that folder
- The script then uses abTEM to generate a simulated 4DSTEM data set from the input structure
- The script saves the data to file
- The process repeats itself until a 4DSTEM data set is generated for all structure files in the folder. 

I run these experiments on an Nvidia GPU. I found that my GPU's memory would quickly become full after a few cycles of the loop, where the number of cycles depended on the simulation parameters. I tested each input structure separately with the same parameters and my GPU was capable of handling each simulation. So, I surmised that my GPU's memory was not being cleared properly after each iteration of loop. For calculations on Nvidia GPUs, abTEM uses CuPy and CUDA. I tested some of the code shown under the heading 'Memory Pool Operations' at https://docs.cupy.dev/en/stable/user_guide/memory.html in an attempt to release memory. My GPU's memory was still not clearing after each iteration of the loop. I was using CUDA 10.2 when I first encountered this issue. After upgrading to 11.3, I have not seen any issues with my GPU's memory. **All I need to do is add garbage collection with 'gc.collect()' at the end of each iteration and the memory is freed** (_this did not work on my system with older versions of CUDA/CuPy_). I was able to run 10 identical simulations in a loop and the available memory after garbage collection was approximately the same as it was at the very start of the script execution. I have a 980Ti with 6 GB of VRAM, so these numbers are 5000 MB free at the start (not sure why it's not 6 GB) and 4850 MB free after each iteration. Without gc.collect(), the amount of free VRAM _usually_ decreases with each iteration. 

For those interested I use the code from the accepted answer from https://stackoverflow.com/questions/59567226/how-to-programmatically-determine-available-gpu-memory-with-tensorflow to check my available GPU memory. 

Hope this information is helpful to someone in the future. 

Best wishes.

Gearóid.

**Comment by jacobjma** [comment]
Dear Gearoid,

Thank you very much for the thorough feedback. Your memory issues may have been due to how cupy handled fft plans, or maybe still handle them, these can really blow up your memory, so I often turn them off. Your comment on gc.collect() is interesting, I would be interested if you could run your code without fft plans, and see whether that makes a difference, you need to run this at the start of your notebook/script:
```
from cupy.fft.config import get_plan_cache
cache = get_plan_cache()
cache.set_size(0)
```

In any case, I hope this discussion will be somewhat outdated soon. We will solve this and many other memory-related issues in the new dask version, as this will allow us to use `dask_cuda` to have some easier control of GPU memory.

Cheers,
Jacob

**Comment by GearoidM** [comment]
Hello, 

Thanks for your response. I added the code you posted to the start of my script and found that without gc.collect(), the amount of available memory still goes down with each iteration; however, interestingly, the addition of these lines somehow results in the availability of more memory. For example, I was running my script on a cubic iron nanoparticle and found that 4.7 GB of memory was available after each iteration with gc.collect() with fft plans. When I disabled fft plans with the code you gave, the amount available after each iteration was 4.85 GB. Furthermore, when I also disable the memory pool with the following code, the available memory jumps to 5.1 GB at the cost of about 1 s of extra computation time:

cp.cuda.set_allocator(None)
cp.cuda.set_pinned_memory_allocator(None)

As I previously mentioned, with older versions of CUDA and CuPy I was unable to free memory with the gc.collect() command and have seen a number of posts on CUDA complaining about memory management. Updating CUDA and CuPy seems to have solved it for now, but I am very much looking forward to the Dask-ified version of abTEM, which I hope will enable me to run some of my calculations locally instead of on a supercomputer (they are very large). 

Best wishes.

Gearóid.
